kernel/printf.c

```diff
@@ -0,0 +1,249 @@
+#include "printf.h"
+
+#include "com.h"
+#include "util.h"
+
+#include <stdarg.h>
+#include <stdint.h>
+
+#define emit(c) do { com_putch(c); ++ret; } while(0)
+
+#define CONV_LEFT_JUST 0x01
+#define CONV_USE_PLUS  0x02
+#define CONV_USE_SPACE 0x04
+#define CONV_ZERO_PAD  0x08
+
+enum arg_size {
+  SIZE_SHORT,
+  SIZE_DEFAULT,
+  SIZE_LONG,
+};
+
+int printf (const char *fmt, ...) {
+  va_list argp;
+  va_start(argp, fmt);
+  int ret = 0;
+  int i = 0;
+
+  while (fmt[i]) {
+    char c = fmt[i++];
+    if (c != '%') {
+      emit(c);
+      continue;
+    }
+    char flags = 0;
+    enum arg_size sz = SIZE_DEFAULT;
+    uint64_t min_width = 0;
+    c = fmt[i++];
+    while (c) {
+      if (c == '-') {
+        flags |= CONV_LEFT_JUST;
+      } else if (c == '+') {
+        flags |= CONV_USE_PLUS;
+      } else if (c == '0') {
+        flags |= CONV_ZERO_PAD;
+      } else if (c == ' ') {
+        flags |= CONV_USE_SPACE;
+      } else {
+        break;
+      }
+      c = fmt[i++];
+    }
+
+    while (c) {
+      if ('0' <= c && c <= '9') {
+        min_width = min_width * 10 + (c - '0');
+        c = fmt[i++];
+      } else {
+        break;
+      }
+    }
+
+    uint8_t bits = 32;
+    if (c == 'h') {
+      sz = SIZE_SHORT;
+      bits = 16;
+      c = fmt[i++];
+    } else if (c == 'l') {
```

Comment:
we don't support floats at *all* (because they're not, like, enabled in the kernel?) so I didn't really wory about it